File: examples/typescript/case02.ts
// Synthetic protected Job Store: each method is one atomic mutation.
export class JobStore {
  currentGeneration = 7;
  state = "RUNNING";
  result: string | null = null;

  takeover(generation: number): void {
    this.currentGeneration = generation;
  }

  // faultscope:begin fs-c02.commit-without-generation
  storeResultWeak(result: string): boolean {
    if (this.state !== "RUNNING") return false;
    this.result = result; // A7 can commit even after B8 takes over.
    return true;
  }
  // faultscope:end fs-c02.commit-without-generation

  // faultscope:begin fs-c02.commit-with-generation
  storeResultGated(acquiredGeneration: number, result: string): boolean {
    // Compare and write atomically at the protected resource.
    if (this.state !== "RUNNING" || acquiredGeneration !== this.currentGeneration) {
      return false;
    }
    this.result = result;
    return true;
  }
  // faultscope:end fs-c02.commit-with-generation
}

// faultscope:begin fs-c02.local-authority-check
export function localCheckThenWeakCommit(
  leaseValid: boolean,
  store: JobStore,
  result: string,
): boolean {
  if (!leaseValid) return false;
  // A takeover can happen after this local check.
  return store.storeResultWeak(result);
}
// faultscope:end fs-c02.local-authority-check

// faultscope:begin fs-c02.current-generation-commit
export function commitCurrentWorker(store: JobStore, result: string): boolean {
  return store.storeResultGated(8, result); // B8 still makes progress.
}
// faultscope:end fs-c02.current-generation-commit

File: examples/cpp/case02.cpp
#include <string>

// Synthetic protected Job Store; each method models one atomic mutation.
struct Case02JobStore {
    int current_generation = 7;
    bool running = true;
    std::string result;

    void takeover(int generation) { current_generation = generation; }

    // faultscope:begin fs-c02.commit-without-generation
    bool store_result_weak(const std::string& value) {
        if (!running) return false;
        result = value; // A7 can commit after B8 takes over.
        return true;
    }
    // faultscope:end fs-c02.commit-without-generation

    // faultscope:begin fs-c02.commit-with-generation
    bool store_result_gated(int acquired_generation, const std::string& value) {
        // Check and write atomically at the protected Job Store.
        if (!running || acquired_generation != current_generation) return false;
        result = value;
        return true;
    }
    // faultscope:end fs-c02.commit-with-generation
};

// faultscope:begin fs-c02.local-authority-check
bool case02_local_check_then_weak_commit(bool lease_valid, Case02JobStore& store, const std::string& value) {
    if (!lease_valid) return false;
    // A takeover may occur after this local check.
    return store.store_result_weak(value);
}
// faultscope:end fs-c02.local-authority-check

// faultscope:begin fs-c02.current-generation-commit
bool case02_commit_current_worker(Case02JobStore& store, const std::string& value) {
    return store.store_result_gated(8, value); // B8 still makes progress.
}
// faultscope:end fs-c02.current-generation-commit
